### PR TITLE
St/update portfolio outfits

### PR DESCRIPTION
**Pull request by saurabhbakolia**
update portfolio outfits method added

### CODE REVIEW

File: src/main/java/com/darzee/shankh/service/PortfolioService.java [hunk 1]
```diff
@@ -257,6 +257,109 @@ public ResponseEntity<CreatePortfolioOutfitResponse> createPortfolioOutfits(Crea
...
+                portfolioOutfitDao.setOutfitType(outfitType);
+            }
+            if (request.getSubOutfit() != null) {
+                int subOutfitType = request.getSubOutfit();
+                outfitType = OutfitType.getOutfitOrdinalEnumMap().get(request.getOutfitType());
+                Set<Integer> possibleSubOutfits = outfitService.getSubOutfitMap(outfitType).keySet();
+                if (!possibleSubOutfits.contains(subOutfitType)) {
+                    throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Invalid Sub Outfit Type!");
+                }
+                portfolioOutfitDao.setSubOutfitType(request.getSubOutfit());
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/PortfolioService.java:282]: this is duplicate code. Create a new method, say 
getSubOutfit(Integer outfitTypeOrdinal, Integer subOutfitOrdinal) , and call it from wherever required.

File: src/main/java/com/darzee/shankh/service/PortfolioService.java [hunk 2]
```diff
@@ -257,6 +257,109 @@ public ResponseEntity<CreatePortfolioOutfitResponse> createPortfolioOutfits(Crea
...
+                                                     Long portfolioOutfitId) {
+        List<String> portfolioOutfitReferenceIds = request.getReferenceIds();
+        if (!CollectionUtils.isEmpty(portfolioOutfitReferenceIds)) {
+            List<String> existingReferenceIds = objectImagesService.getPortfolioOutfitsReferenceIds(portfolioOutfitId);
+            if (CollectionUtils.isEmpty(existingReferenceIds)) {
+                objectImagesService.saveObjectImages(portfolioOutfitReferenceIds,
+                        ImageEntityType.PORTFOLIO_OUTFIT.getEntityType(),
+                        portfolioOutfitId);
+            }
+            if (!CollectionUtils.isEmpty(existingReferenceIds)) {
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/PortfolioService.java:319]: No need of  if (!CollectionUtils.isEmpty(existingReferenceIds)) {

instead use
if(CollectionUtils.isEmpty(existingReferenceIds)) {
//LOGIC
return;
}
//LOGIC

File: src/main/java/com/darzee/shankh/service/PortfolioService.java [hunk 3]
```diff
@@ -257,6 +257,109 @@ public ResponseEntity<CreatePortfolioOutfitResponse> createPortfolioOutfits(Crea
...
+                List<String> removedReferenceIds = existingReferenceIds
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/PortfolioService.java:320]: no need of stream here . Instead use something like
List<Integer> c = new ArrayList<>(a);
c.removeAll(b);

File: src/main/java/com/darzee/shankh/service/PortfolioService.java [hunk 4]
```diff
@@ -257,6 +257,109 @@ public ResponseEntity<CreatePortfolioOutfitResponse> createPortfolioOutfits(Crea
...
+                        .stream()
+                        .filter(exisingId -> !portfolioOutfitReferenceIds.contains(exisingId))
+                        .collect(Collectors.toList());
+
+                List<String> updatedReferenceIds = portfolioOutfitReferenceIds
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/PortfolioService.java:325]: same as above

File: src/main/java/com/darzee/shankh/service/PortfolioService.java [hunk 5]
```diff
@@ -257,6 +257,109 @@ public ResponseEntity<CreatePortfolioOutfitResponse> createPortfolioOutfits(Crea
...
+                        .stream()
+                        .filter(newId -> !existingReferenceIds.contains(newId))
+                        .collect(Collectors.toList());
+
+                for (String removedId : removedReferenceIds) {
+                    objectImagesService.invalidateExistingReferenceId(removedId);
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/PortfolioService.java:331]: create a new method
invalidateExistingReferenceIds(List<String> invalidReferenceIds) for bulk support

File: src/main/java/com/darzee/shankh/service/PortfolioService.java [hunk 6]
```diff
@@ -257,6 +257,109 @@ public ResponseEntity<CreatePortfolioOutfitResponse> createPortfolioOutfits(Crea
...
+            return new ResponseEntity<>(response, HttpStatus.OK);
+
+        }
+        response.setMessage("Portfolio outfit not found!");
+        return new ResponseEntity<>(response, HttpStatus.OK);
+    }
+
+    private void updatePortfolioOutfitImageReference(CreatePortfolioOutfitRequest request,
+                                                     PortfolioOutfitsDAO updatePortfolioOutfit,
+                                                     Long portfolioOutfitId) {
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/PortfolioService.java:310]: portfolioId is not required as a seperate param here. fetch it from updatePortfolioOutfit wherever required

File: src/main/java/com/darzee/shankh/service/PortfolioService.java [hunk 7]
```diff
@@ -257,6 +257,109 @@ public ResponseEntity<CreatePortfolioOutfitResponse> createPortfolioOutfits(Crea
...
+                }
+                if (!CollectionUtils.isEmpty(updatedReferenceIds)) {
+                    objectImagesService.saveObjectImages(updatedReferenceIds, ImageEntityType.PORTFOLIO_OUTFIT.getEntityType(),
+                            portfolioOutfitId);
+                }
+            }
+        }
+    }
+
+    public ResponseEntity<CreatePortfolioOutfitResponse> deletePortfolioOutfit(Long portfolioOutfitId) throws Exception {
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/PortfolioService.java:341]: rename this to invalidatePortfolioOutfit

File: src/main/java/com/darzee/shankh/controller/PortfolioController.java [hunk 8]
```diff
@@ -35,6 +35,17 @@ public ResponseEntity<CreatePortfolioOutfitResponse> createPortfolioOutfit(@Requ
...
 
+    @PutMapping(value = "/portfolio_outfit/{portfolio_outfit_id}", produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<CreatePortfolioOutfitResponse> updatePortfolioOutfit(@RequestBody @Valid CreatePortfolioOutfitRequest request,
+                                                                             @PathVariable("portfolio_outfit_id") Long portfolioOutfitId) throws Exception {
+        return portfolioService.updatePortfolioOutfits(request, portfolioOutfitId);
+    }
+
+    @PutMapping(value = "/delete_portfolio_outfit/{portfolio_outfit_id}", produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<CreatePortfolioOutfitResponse> deletePortfolioOutfit(@PathVariable("portfolio_outfit_id") Long portfolioOutfitId) throws Exception {
+        return portfolioService.deletePortfolioOutfit(portfolioOutfitId);
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/controller/PortfolioController.java:46]: replace delete with invalidate. check in url as well.

File: src/main/java/com/darzee/shankh/service/PortfolioService.java [hunk 9]
```diff
@@ -257,6 +257,109 @@ public ResponseEntity<CreatePortfolioOutfitResponse> createPortfolioOutfits(Crea
...
+        CreatePortfolioOutfitResponse response = new CreatePortfolioOutfitResponse();
+        Optional<PortfolioOutfits> portfolioOutfit = portfolioOutfitsRepo.findById(portfolioOutfitId);
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/PortfolioService.java:343]: always use, findByIdAndIsValid(Long id, Boolean isValid)

File: src/main/java/com/darzee/shankh/service/PortfolioService.java [hunk 10]
```diff
@@ -257,6 +257,109 @@ public ResponseEntity<CreatePortfolioOutfitResponse> createPortfolioOutfits(Crea
         return new ResponseEntity<>(response, HttpStatus.CREATED);
     }
 
+    public ResponseEntity<CreatePortfolioOutfitResponse> updatePortfolioOutfits(CreatePortfolioOutfitRequest request,
+                                                                                Long portfolioOutfitId) throws Exception {
+        CreatePortfolioOutfitResponse response = new CreatePortfolioOutfitResponse();
+        Optional<PortfolioOutfits> portfolioOutfit = portfolioOutfitsRepo.findById(portfolioOutfitId);
```
[the remark this answers is not in the document]
Comment (saurabhbakolia) [reply src/main/java/com/darzee/shankh/service/PortfolioService.java:263]: Added this check.
(portfolioOutfit.isPresent() && portfolioOutfit.get().getIsValid())

File: src/main/java/com/darzee/shankh/service/ObjectImagesService.java [hunk 11]
```diff
@@ -133,6 +134,18 @@ public void invalidateExistingReferenceIds(String entityType, Long entityId) {
         }
     }
 
+    public void invalidateExistingReferenceIds(List<String> invalidReferenceIds){
+        for (String referenceId : invalidReferenceIds){
+            Optional<ObjectImages> validImage = repo.findAllByReferenceIdAndIsValid(referenceId,
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/ObjectImagesService.java:139]: what if there exist multiple images with same reference id?

Comment (saurabhbakolia) [reply src/main/java/com/darzee/shankh/service/ObjectImagesService.java:139]: We can add another loop then. 
for(ObjectImages validImage : validImages) {
//Logic
}

**Pull request merged**